### PR TITLE
Automatic port of System.Threading exceptions

### PR DESCRIPTION
**Pull request by carlossanlop**
Ported using [DocsPortingTool ](https://github.com/carlossanlop/DocsPortingTool/) with the `-PortMemberExceptions true` argument.

### CODE REVIEW

File: xml/System.Threading.Tasks/TaskCompletionSource.xml [hunk 1]
```diff
@@ -117,6 +118,7 @@ The <xref:System.Threading.Tasks.Task> created by this instance and accessible t
         <param name="creationOptions">The options to use when creating the underlying <see cref="T:System.Threading.Tasks.Task" />.</param>
         <summary>Creates a <see cref="T:System.Threading.Tasks.TaskCompletionSource" /> with the specified state and options.</summary>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException">The <paramref name="creationOptions" /> represent options invalid for use with a <see cref="T:System.Threading.Tasks.TaskCompletionSource" />.</exception>
```
Comment (gewarren) [inline xml/System.Threading.Tasks/TaskCompletionSource.xml:121]:
```suggestion
        <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="creationOptions" /> represents options that are invalid for use with a <see cref="T:System.Threading.Tasks.TaskCompletionSource" />.</exception>
```

File: xml/System.Threading.Tasks/TaskCompletionSource.xml [hunk 2]
```diff
@@ -186,6 +190,10 @@ The <xref:System.Threading.Tasks.Task> created by this instance and accessible t
         <param name="exceptions">The collection of exceptions to bind to this <see cref="T:System.Threading.Tasks.Task" />.</param>
         <summary>Transitions the underlying <see cref="T:System.Threading.Tasks.Task" /> into the <see cref="F:System.Threading.Tasks.TaskStatus.Faulted" /> state.</summary>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentNullException">
+          <paramref name="exceptions" /> is <see langword="null" />.</exception>
+        <exception cref="T:System.ArgumentException">There are one or more <see langword="null" /> elements in <paramref name="exceptions" />.</exception>
```
Comment (gewarren) [inline xml/System.Threading.Tasks/TaskCompletionSource.xml:195]:
```suggestion
        <exception cref="T:System.ArgumentException">There is one or more <see langword="null" /> elements in <paramref name="exceptions" />.</exception>
```

File: xml/System.Threading/Interlocked.xml [hunk 3]
```diff
@@ -1181,6 +1189,7 @@ If `comparand` and the object in `location1` are equal by reference, then `value
         <summary>Decrements a specified variable and stores the result, as an atomic operation.</summary>
         <returns>The decremented value.</returns>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.NullReferenceException">The address of location is a <see langword="null" /> pointer.</exception>
```
Comment (gewarren) [inline xml/System.Threading/Interlocked.xml:1192]:
```suggestion
        <exception cref="T:System.NullReferenceException">The address of <paramref name="location" /> is a <see langword="null" /> pointer.</exception>
```

File: xml/System.Threading/Interlocked.xml [hunk 4]
```diff
@@ -1212,6 +1221,7 @@ If `comparand` and the object in `location1` are equal by reference, then `value
...
```
Comment (gewarren) [inline xml/System.Threading/Interlocked.xml:1224]:
```suggestion
        <exception cref="T:System.NullReferenceException">The address of <paramref name="location" /> is a <see langword="null" /> pointer.</exception>
```

File: xml/System.Threading/Interlocked.xml [hunk 5]
```diff
@@ -1603,6 +1613,7 @@ If `comparand` and the object in `location1` are equal by reference, then `value
         <summary>Sets a 32-bit unsigned integer to a specified value and returns the original value, as an atomic operation.</summary>
         <returns>The original value of <paramref name="location1" />.</returns>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.NullReferenceException">The address of location1 is a <see langword="null" /> pointer.</exception>
```
Comment (gewarren) [inline xml/System.Threading/Interlocked.xml:1616]:
```suggestion
        <exception cref="T:System.NullReferenceException">The address of <paramref name="location1" /> is a <see langword="null" /> pointer.</exception>
```

File: xml/System.Threading/Interlocked.xml [hunk 6]
```diff
@@ -1636,6 +1647,7 @@ If `comparand` and the object in `location1` are equal by reference, then `value
         <summary>Sets a 64-bit unsigned integer to a specified value and returns the original value, as an atomic operation.</summary>
         <returns>The original value of <paramref name="location1" />.</returns>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.NullReferenceException">The address of location1 is a <see langword="null" /> pointer.</exception>
```
Comment (gewarren) [inline xml/System.Threading/Interlocked.xml:1650]:
```suggestion
        <exception cref="T:System.NullReferenceException">The address of <paramref name="location1" /> is a <see langword="null" /> pointer.</exception>
```

File: xml/System.Threading/Interlocked.xml [hunk 7]
```diff
@@ -1893,6 +1905,7 @@ If `comparand` and the object in `location1` are equal by reference, then `value
         <summary>Increments a specified variable and stores the result, as an atomic operation.</summary>
         <returns>The incremented value.</returns>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.NullReferenceException">The address of location is a <see langword="null" /> pointer.</exception>
```
Comment (gewarren) [inline xml/System.Threading/Interlocked.xml:1908]:
```suggestion
        <exception cref="T:System.NullReferenceException">The address of <paramref name="location" /> is a <see langword="null" /> pointer.</exception>
```

File: xml/System.Threading/Interlocked.xml [hunk 8]
```diff
@@ -1924,6 +1937,7 @@ If `comparand` and the object in `location1` are equal by reference, then `value
...
```
Comment (gewarren) [inline xml/System.Threading/Interlocked.xml:1940]:
```suggestion
        <exception cref="T:System.NullReferenceException">The address of <paramref name="location" /> is a <see langword="null" /> pointer.</exception>
```

File: xml/System.Threading.Tasks/TaskCompletionSource.xml (hunk 1 above)
Comment (kouvel) [reply xml/System.Threading.Tasks/TaskCompletionSource.xml:121]: Maybe:
```xml
One or more of the specified <paramref name="creationOptions" /> are invalid for use with a <see cref="T:System.Threading.Tasks.TaskCompletionSource" />.
```

If a change is made the same change should be made above as well.

File: xml/System.Threading/CountdownEvent.xml [hunk 9]
```diff
@@ -745,7 +745,10 @@
           <paramref name="signalCount" /> is less than or equal to 0.</exception>
         <exception cref="T:System.InvalidOperationException">
           <see cref="P:System.Threading.CountdownEvent.CurrentCount" /> + <paramref name="signalCount" /> is equal to or greater than <see cref="F:System.Int32.MaxValue" />.
-        </exception>
+
+-or-
+
+The current instance is already set.</exception>
```
Comment (kouvel) [inline xml/System.Threading/CountdownEvent.xml:751]: This comment applies to `AddCount`, which throws when the event is signaled. `TryAddCount` just returns false, so this doesn't apply here.

File: xml/System.Threading/EventWaitHandleAcl.xml [hunk 10]
```diff
@@ -67,6 +67,7 @@ If a `name` is passed and the system event already exists, the existing event is
         <exception cref="T:System.IO.DirectoryNotFoundException">Could not find a part of the path specified in <paramref name="name" />.</exception>
         <exception cref="T:System.Threading.WaitHandleCannotBeOpenedException">A system-wide synchronization event with the provided <paramref name="name" /> was not found.
  -or- An <see cref="T:System.Threading.EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="T:System.Threading.EventWaitHandle" /> of a different type might have the same name.</exception>
+        <exception cref="T:System.ArgumentNullException">.NET Framework only: The <paramref name="name" /> length is beyond MAX_PATH (260 characters).</exception>
```
Comment (kouvel) [inline xml/System.Threading/EventWaitHandleAcl.xml:70]: .NET Framework actually throws `ArgumentException` in this case

File: xml/System.Threading.Tasks/TaskCompletionSource.xml (hunk 1 above)
Comment (carlossanlop) [reply xml/System.Threading.Tasks/TaskCompletionSource.xml:121]: I like @kouvel 's suggestion, it sounds similar to how we have worded similar exceptions for enums out of range in other places.

```suggestion
        <exception cref="T:System.ArgumentOutOfRangeException">One or more of the specified <paramref name="creationOptions" /> are invalid for use with a <see cref="T:System.Threading.Tasks.TaskCompletionSource" />.
</exception>
```

File: xml/System.Threading.Tasks/TaskCompletionSource.xml [hunk 11]
```diff
@@ -94,6 +94,7 @@ The <xref:System.Threading.Tasks.Task> created by this instance and accessible t
 
           ]]></format>
         </remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException">The <paramref name="creationOptions" /> represent options invalid for use with a <see cref="T:System.Threading.Tasks.TaskCompletionSource" />.</exception>
```
Comment (carlossanlop) [inline xml/System.Threading.Tasks/TaskCompletionSource.xml:97]:
```suggestion
        <exception cref="T:System.ArgumentOutOfRangeException">One or more of the specified <paramref name="creationOptions" /> are invalid for use with a <see cref="T:System.Threading.Tasks.TaskCompletionSource" />.
</exception>
```

File: xml/System.Threading/CountdownEvent.xml (hunk 9 above)
Comment (carlossanlop) [inline xml/System.Threading/CountdownEvent.xml:751]:
```suggestion
          </exception>
```

File: xml/System.Threading/EventWaitHandleAcl.xml (hunk 10 above)
Comment (carlossanlop) [reply xml/System.Threading/EventWaitHandleAcl.xml:70]:
```suggestion
        <exception cref="T:System.ArgumentException">.NET Framework only: The <paramref name="name" /> length is beyond MAX_PATH (260 characters).</exception>
```

File: xml/System.Threading/CountdownEvent.xml (hunk 9 above)
Comment (carlossanlop) [reply xml/System.Threading/CountdownEvent.xml]: Removed.

File: xml/System.Threading.Tasks/TaskCompletionSource.xml [hunk 12]
```diff
@@ -192,7 +194,7 @@ The <xref:System.Threading.Tasks.Task> created by this instance and accessible t
         <remarks>To be added.</remarks>
         <exception cref="T:System.ArgumentNullException">
           <paramref name="exceptions" /> is <see langword="null" />.</exception>
-        <exception cref="T:System.ArgumentException">There are one or more <see langword="null" /> elements in <paramref name="exceptions" />.</exception>
+        <exception cref="T:System.ArgumentException">There is one or more <see langword="null" /> elements in <paramref name="exceptions" />.</exception>
```
Comment (kouvel) [inline xml/System.Threading.Tasks/TaskCompletionSource.xml:197]: I think `are` may be correct, currently it reads "There is (x) elements", I think it is more correct to say "There are (x) elements". As for "There is one or more (x)" vs "There are one more (x)" I'm not sure but even then the latter sounds more correct to me and (x) then must be plural.

Comment (gewarren) [reply xml/System.Threading.Tasks/TaskCompletionSource.xml:197]: Based on what I can tell from https://english.stackexchange.com/questions/13284/which-is-correct-one-or-more-is-or-one-or-more-are, it should be:

1) There is one or more...

-and-

2) One or more are...

Based on proximity of the noun to the verb. In the first example, "one" is closer to the verb. In the second example, "more" is closer to the verb.

Comment (gewarren) [reply xml/System.Threading.Tasks/TaskCompletionSource.xml:197]: @mairaw Any thoughts?

Comment (kouvel) [reply xml/System.Threading.Tasks/TaskCompletionSource.xml:197]: > https://english.stackexchange.com/questions/13284/which-is-correct-one-or-more-is-or-one-or-more-are
>  the verb should be singular, as long as the things in the list are individually singular

Based on this it sounds to me like "one or more" would be a list of one singular and one plural term. I haven't read into it much though.

Comment (carlossanlop) [reply xml/System.Threading.Tasks/TaskCompletionSource.xml:197]:
```suggestion
        <exception cref="T:System.ArgumentException">There are one or more <see langword="null" /> elements in <paramref name="exceptions" />.</exception>
```

**Pull request merged**